src/target/riscv/riscv-013.c

```diff
@@ -3396,310 +3396,524 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
+		struct memory_access_info access)
+{
+	LOG_TARGET_DEBUG(target, "DMI_STATUS_BUSY encountered in batch. Memory read [%"
+			PRIu32 ", %" PRIu32 ")", start_index, next_start_index);
+	if (start_index == next_start_index)
+		LOG_TARGET_DEBUG(target, "No forward progress");
 
-		/* Wait for the target to finish performing the last abstract command,
-		 * and update our copy of cmderr. If we see that DMI is busy here,
-		 * dmi_busy_delay will be incremented. */
-		uint32_t abstractcs;
-		if (dmi_read(target, &abstractcs, DM_ABSTRACTCS) != ERROR_OK)
+	if (dmi_write(target, DM_ABSTRACTAUTO, 0) != ERROR_OK)
+		return ERROR_FAIL;
+	return read_memory_progbuf_inner_startup(target, access.target_address,
+			access.increment, next_start_index);
+}
+
+/**
+ * This function extracts the data from the batch.
+ */
+static int read_memory_progbuf_inner_extract_batch_data(struct target *target,
+		const struct riscv_batch *batch,
+		uint32_t start_index, uint32_t elements_to_read, uint32_t *elements_read,
+		struct memory_access_info access)
+{
+	const bool two_reads_per_element = access.element_size > 4;
+	const uint32_t reads_per_element = (two_reads_per_element ? 2 : 1);
+	assert(!two_reads_per_element || riscv_xlen(target) == 64);
+	assert(elements_to_read <= UINT32_MAX / reads_per_element);
+	const uint32_t nreads = elements_to_read * reads_per_element;
+	for (uint32_t curr_idx = start_index, read = 0; read < nreads; ++read) {
+		switch (riscv_batch_get_dmi_read_op(batch, read)) {
+		case DMI_STATUS_BUSY:
+			*elements_read = curr_idx - start_index;
+			return read_memory_progbuf_inner_on_dmi_busy(target, start_index, curr_idx
+					, access);
+		case DMI_STATUS_FAILED:
+			LOG_TARGET_DEBUG(target,
+					"Batch memory read encountered DMI_STATUS_FAILED on read %"
+					PRIu32, read);
 			return ERROR_FAIL;
-		while (get_field(abstractcs, DM_ABSTRACTCS_BUSY))
-			if (dmi_read(target, &abstractcs, DM_ABSTRACTCS) != ERROR_OK)
-				return ERROR_FAIL;
-		info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
+		case DMI_STATUS_SUCCESS:
+			break;
+		default:
+			assert(0);
+		}
+		const uint32_t value = riscv_batch_get_dmi_read_data(batch, read);
+		uint8_t * const curr_buff = access.buffer_address +
+			curr_idx * access.element_size;
+		const target_addr_t curr_addr = access.target_address +
+			curr_idx * access.increment;
+		const uint32_t size = access.element_size;
+
+		assert(size <= 8);
+		const bool is_odd_read = read % 2;
+
+		if (two_reads_per_element && !is_odd_read) {
+			buf_set_u32(curr_buff + 4, 0, (size * 8) - 32, value);
+			continue;
+		}
+		const bool is_second_read = two_reads_per_element;
 
-		unsigned next_index;
-		unsigned ignore_last = 0;
-		switch (info->cmderr) {
-			case CMDERR_NONE:
-				LOG_DEBUG("successful (partial?) memory read");
-				next_index = index + reads;
-				break;
-			case CMDERR_BUSY:
-				LOG_DEBUG("memory read resulted in busy response");
+		buf_set_u32(curr_buff, 0, is_second_read ? 32 : size * 8, value);
```

Comment:
Minor thing - for readability, you can put here extra parentheses:

```suggestion
		buf_set_u32(curr_buff, 0, is_second_read ? 32 : (size * 8), value);
```

Reply:
Addressed.